x/superfluid/keeper/stake_test.go

```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+			minAmtToStake := sdk.ZeroInt()
+			if tc.useMinAmountToStake {
+				minAmtToStake = sdk.NewInt(9999999999)
+			}
+			sharesToStake := shareOut
+			if tc.usePartialShares {
+				sharesToStake.Amount = sharesToStake.Amount.Quo(sdk.NewInt(2))
+			}
+
+			balanceBeforeConvert := s.App.BankKeeper.GetBalance(s.Ctx, sender, shareOut.Denom)
+			s.Require().True(!balanceBeforeConvert.Amount.IsZero())
+
+			bondDenom := s.App.StakingKeeper.BondDenom(s.Ctx)
+			totalPoolLiquidityBeforeConvert, err := s.App.GAMMKeeper.GetTotalPoolLiquidity(s.Ctx, poolId)
+			s.Require().NoError(err)
+			bondDenomPoolAmtBeforeConvert := totalPoolLiquidityBeforeConvert.AmountOf(bondDenom)
+
+			// system under test
+			totalAmtConverted, sharesDelegated, err := s.App.SuperfluidKeeper.ConvertUnlockedToStake(s.Ctx, sender, valAddr.String(), shareOut, minAmtToStake)
+			if tc.expectedError {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+
+			// gamm check
+			totalPoolLiquidityAfterConvert, err := s.App.GAMMKeeper.GetTotalPoolLiquidity(s.Ctx, poolId)
+			s.Require().NoError(err)
+			// check that pool liquidity have reduced
+			bondDenomPoolAmtAfterConvert := totalPoolLiquidityAfterConvert.AmountOf(bondDenom)
+			s.Require().True(bondDenomPoolAmtAfterConvert.LT(bondDenomPoolAmtBeforeConvert))
+
+			// Staking & Delegation check
+			// check if delegation amount matches
+			delegation, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, valAddr)
+			s.Require().True(found)
+			s.Require().Equal(delegation.Shares, sharesDelegated)
+			s.Require().True(totalAmtConverted.ToDec().Equal(delegation.Shares))
+
+			// Bank check
+			balanceAfterConvertLockToStake := s.App.BankKeeper.GetBalance(s.Ctx, sender, shareOut.Denom)
+			s.Require().True(balanceAfterConvertLockToStake.IsZero())
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestValidateUnbondConvertAndStake() {
+	type tc struct {
+		sharesToMigrate sdk.Coin
+		expectedError   bool
+	}
+	testCases := map[string]tc{
+		"happy case": {
+			sharesToMigrate: sdk.NewInt64Coin("gamm/pool/1", 100),
+		},
+		"non gamm token prefix": {
+			sharesToMigrate: sdk.NewInt64Coin("uosmo", 100),
+			expectedError:   true,
+		},
+		"invalid gamm token": {
+			sharesToMigrate: sdk.NewInt64Coin("gamm/pool/invalid", 100),
+			expectedError:   true,
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+			s.SetupTest()
+
+			poolIdLeaving, err := s.App.SuperfluidKeeper.ValidateUnbondConvertAndStake(s.Ctx, tc.sharesToMigrate)
+			if tc.expectedError {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+			s.Require().Equal(poolIdLeaving, uint64(1))
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestConvertGammSharesToOsmoAndStake() {
+	type tc struct {
+		useInvalidValAddr        bool
+		useMinAmtToStake         bool
+		useValSetPrefSingleVal   bool
+		useValSetPrefMultipleVal bool
+
+		expectedError bool
+	}
+	testCases := map[string]tc{
+		"happy case": {},
```

Comment:
can we define what the happy case actually is? I would argue there is no happy case, just multiple cases that test each branch of this method